File: c/src/pin.c
#include "pin.h"

#include "bits.h"
#include "test_io.h"
#include "payments.h"
#include "crypto.h"

#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#include <openssl/crypto.h>
#include <openssl/rand.h>
#include <openssl/aes.h>
#include <openssl/des.h>

static size_t pin_block_size[] = {TDES_BLOCK_SIZE, TDES_BLOCK_SIZE, TDES_BLOCK_SIZE, TDES_BLOCK_SIZE, 2*AES_BLOCK_SIZE};

/* Returns the PIN block size, or 0 if the format is invalid.
 * For format 4, the size is doubled due to the need to perform a CBC encryption.
 * @param format the format code
 * @result PIN block size in bytes
 */
size_t get_pin_block_size(int format){
	if (!VALID_PIN_BLOCK_FORMAT(format)) return 0;
	return pin_block_size [format];
}

/**
 * Prepares a pin block of the given format. The PIN block is written in a packed form
 * into the output variable.
 *
 * In case of PIN block format 4, two separate blocks are returned (they will have to be CBC encrypted)
 * @param format PIN block format, 0 to 4
 * @pin	PIN value, unpacked BCD
 * @pin_len PIN length
 * @pan PAN value, unpacked BCD
 * @pan_len PAN length
 * @unique_id unique transaction ID PIN block format 1. If NULL, ignored for format 1.
 * @unique_id_len length of the unique ID
 * @output pointer to the output buffer which must be of MAX_PIN_BLOCK_SIZE
 * @result 0 if successful.
 */
int make_pin_block (int format, uint8_t * pin, size_t pin_len, uint8_t * pan, size_t pan_len,
		uint8_t * unique_id, size_t unique_id_len, uint8_t * output) {
	/* validate the input */
	if (!VALID_PIN_BLOCK_FORMAT(format)) return PIN_ERROR;

	if (PIN_BLOCK_FORMAT_1!=format && PIN_BLOCK_FORMAT_2!=format) {
		if (!(pan && pan_len)) return PIN_ERROR;
	 	if (!VALID_PAN_LENGTH(pan_len)) return PIN_ERROR;
	}

	if (!(pin && pin_len && output)) return PIN_ERROR;

	if (pin_len > MAX_PIN_LENGTH) return PIN_ERROR;

	uint8_t buffer1[2*AES_BLOCK_SIZE], buffer2[2*AES_BLOCK_SIZE];

	memset(buffer1, 0, 2*AES_BLOCK_SIZE);
	memset(buffer2, 0, 2*AES_BLOCK_SIZE);

	size_t offset = 2;
	/* prepare buffer1 */
	buffer1[0] = format;
	buffer1[1] = pin_len;

	memcpy(buffer1+offset, pin, pin_len);
	offset += pin_len;

	/* format 1, if unique ID is provided */
	if (PIN_BLOCK_FORMAT_1==format && unique_id && unique_id_len) {
		size_t len = 2*AES_BLOCK_SIZE - offset;
		len = len>unique_id_len? unique_id_len:len;
		memcpy(buffer1+offset, unique_id, len);
		offset += len;
	}

	/* padding */
	switch (format){
		case PIN_BLOCK_FORMAT_0:
		case PIN_BLOCK_FORMAT_2:
			/* padding is with 0xF */
			memset(buffer1+offset, 0xF, 2*AES_BLOCK_SIZE-offset);
			break;
		case PIN_BLOCK_FORMAT_1:
		case PIN_BLOCK_FORMAT_3:
			/* padding is random */
			RAND_priv_bytes(buffer1+offset,2*AES_BLOCK_SIZE-offset);
			break;
		case PIN_BLOCK_FORMAT_4:
			/* this one is special */
			if (offset<2*TDES_BLOCK_SIZE) {
				memset(buffer1+offset, 0xA, 2*TDES_BLOCK_SIZE);
				offset = 2*TDES_BLOCK_SIZE;
			}
			RAND_priv_bytes(buffer1+offset, 2*AES_BLOCK_SIZE-offset);
			break;
	}
	print_array("\tBuffer 1: ", buffer1, 2*AES_BLOCK_SIZE, "\n");

	if (PIN_BLOCK_FORMAT_4!=format) {
		/* PIN block formats 0-3 can be made via XOR with the second block, which will
		 * simply have to be all zeros if the format is 1 or 2.
		 */

		/* If the format isn't 2, PAN is mandatory */
		if (PIN_BLOCK_FORMAT_1!=format && PIN_BLOCK_FORMAT_2!=format) {
			/* copy the last 12 digits of the PAN without the check digit */
			memcpy(buffer2+4, pan+(pan_len-13), PAN_BLOCK_LEN_0123);
		}
	} else {
		/* second block of format 4 has the structure LPPPPPPPPP0000, where
		 * L is the PAN length-12, PPPP are the PAN digits and the rest is the zero-
		 * padding.
		 */
		buffer2[0] = pan_len-12;
		memcpy(buffer2+1, pan, pan_len);
	}
	print_array("\tBuffer 2: ", buffer2, 32, "\n");

	if (PIN_BLOCK_FORMAT_4 == format) {
		/* simply pack the two buffers into the output */
		pack_bcd(buffer1, 2*AES_BLOCK_SIZE, output, AES_BLOCK_SIZE, PAD_RIGHT);
		pack_bcd(buffer2, 2*AES_BLOCK_SIZE, output+AES_BLOCK_SIZE, AES_BLOCK_SIZE, PAD_RIGHT);
		print_array("\t\tBlock 1: ", output, AES_BLOCK_SIZE, "\n");
		print_array("\t\tBlock 2: ", output+AES_BLOCK_SIZE, AES_BLOCK_SIZE, "\n");
	}
	else {
		uint8_t block1[8], block2[8];
		pack_bcd(buffer1, 2*TDES_BLOCK_SIZE, block1, TDES_BLOCK_SIZE, PAD_RIGHT);
		pack_bcd(buffer2, 2*TDES_BLOCK_SIZE, block2, TDES_BLOCK_SIZE, PAD_RIGHT);
		print_array("\t\tBlock 1: ", block1, TDES_BLOCK_SIZE, "\n");
		print_array("\t\tBlock 2: ", block2, TDES_BLOCK_SIZE, "\n");
		xor_array(block1, block2, output, TDES_BLOCK_SIZE);
	}

	return PIN_OK;
}
/**
 * Encrypts a format 4 block formerly prepared by make_pin_block.
 * @param key AES key
 * @param key_size AES key size, bytes
 * @param input the two parts of the pin block in a single array
 * @param output the output buffer, must be of at least 16 bytes long
 * @result PIN_OK if ok, PIN_ERROR if not.
 */
int encrypt_format_4_block( uint8_t* key, size_t key_size, uint8_t * input, uint8_t* output ) {
	AES_KEY enc_key;
	int result = PIN_ERROR;
	/* The IV for this encryption is all zero*/
	uint8_t iv[AES_BLOCK_SIZE];
	memset(iv, 0, AES_BLOCK_SIZE);

	if (! (key&& key_size && input && output)) return PIN_ERROR;

	if (!VALID_AES_KEY_SIZE(key_size)) return PIN_ERROR;

	/* attempt to initialize the key schedule. From this point onwards, the memory must be cleansed */
	if (AES_set_encrypt_key(key, key_size*8, &enc_key) <0) goto cleanup;

	print_array("\t\tCBC input block: ", input, 2*AES_BLOCK_SIZE, "\n");
	print_array("\t\tCBC IV: ", iv, AES_BLOCK_SIZE, "\n");
	/* perform encryption */
	AES_cbc_encrypt(input, output, 2* AES_BLOCK_SIZE, &enc_key, iv, AES_ENCRYPT);
	print_array("\t\tCBC Output: ", output, AES_BLOCK_SIZE, "\n");

	result = PIN_OK;
cleanup:
	/* clean up the enc_key due to sensitivity */
	PURGE(enc_key);
	return result;
}

/**
 * Decrypts a format 4 block. Returns only the first chunk of it.
 * @param key AES key
 * @param key_size AES key size, bytes
 * @param pan the PAN
 * @param pan_len PAN length
 * @param input the pin block in a single array
 * @param output the output buffer, must be of at least 16 bytes long
 * @result PIN_OK if ok, PIN_ERROR if not.
 */
int decrypt_format_4_block ( uint8_t* key, size_t key_size, uint8_t* pan, size_t pan_len, uint8_t * input, uint8_t* output) {
	int result = PIN_ERROR;
	/* validate the input */
	if (!(key && key_size && pan && pan_len && input && output)) return PIN_ERROR;

	if (!VALID_PAN_LENGTH(pan_len)) return PIN_ERROR;
	if (!VALID_AES_KEY_SIZE(key_size)) return PIN_ERROR;

	AES_KEY dec_key;
	/* prepare for decryption */
	if (AES_set_decrypt_key(key, key_size*8, &dec_key) <0) goto cleanup;

	/* prepare block 2*/
	uint8_t block2_buffer[2*AES_BLOCK_SIZE];
	uint8_t iv[AES_BLOCK_SIZE];

	uint8_t enc_input_buffer[2*AES_BLOCK_SIZE];

	memset( block2_buffer, 0, 2*AES_BLOCK_SIZE);
	memset( iv, 0, AES_BLOCK_SIZE );
	memcpy(enc_input_buffer, input, AES_BLOCK_SIZE);

	block2_buffer[0] = pan_len -12;
	memcpy(block2_buffer+1, pan, pan_len);

	pack_bcd(block2_buffer, 2*AES_BLOCK_SIZE, enc_input_buffer+AES_BLOCK_SIZE, AES_BLOCK_SIZE, PAD_RIGHT);
	/* Decryption. The IV is zeros. The block 2 (constructed from the PAN) is appended to the ciphertext
	 * and the result is decrypted using AES in the CBC mode. */
	AES_cbc_encrypt(enc_input_buffer, output, 2*AES_BLOCK_SIZE, &dec_key, iv, AES_DECRYPT);

	result = PIN_OK;

cleanup:
	PURGE(dec_key);
	return result;
}

/** Encrypts the key under the KEK, applying a variant.
 *  A single-byte variant is assumed for the encryption. The variant array has up to three positions. The kek is
 *  a double TDES key. The input key is a double or a triple TDES key.
 *  @param key the key to encrypt
 *  @param key_len length of the input key to encrypt. It can be either double or triple TDES length.
 *  @kek the key encryption key, always a double TDES key
 *  @variant the variant table, three bytes, applied to the first byte of the second half of the kek
 *  @output the output buffer, must contain the same bytes as key_len
 *  @result PIN_OK on success, PIN_ERROR on failure
 */
int encrypt_key_variant( uint8_t * key, uint8_t key_len, uint8_t *kek, uint8_t* variant, uint8_t* output ) {
	int result = PIN_ERROR;
	/* validate inputs */
	if (!(key && key_len && kek && variant && output )) return PIN_ERROR;

	if (TDES_KEY_LENGTH_2!=key_len & TDES_KEY_LENGTH_3!=key_len) return PIN_ERROR;
 	DES_key_schedule kek_ks[2];

 	/* the first key half is unchanged by this version of variants*/
 	DES_set_key_unchecked((DES_cblock*)kek, &(kek_ks[0]));

 	print_array("\t\tKEK prior to variants: ", kek, TDES_KEY_LENGTH_2, "\n");

 	for (size_t i=0; i< key_len/TDES_KEY_LENGTH_1; i++) {
 		/* apply variant to the first byte of the second half of the key */
 		kek[TDES_KEY_LENGTH_1]^=variant[i];
 		print_array("\t\tKEK after applying the variant: ", kek, TDES_KEY_LENGTH_2, "\n");
 		/* Initiate key schedule */
 		DES_set_key_unchecked((DES_cblock*)kek, &(kek_ks[1]));
 		/* Encrypt the appropriate block */
 		DES_ecb2_encrypt( (DES_cblock*) (key+TDES_BLOCK_SIZE*i), (DES_cblock*) (output+TDES_BLOCK_SIZE*i), kek_ks, kek_ks+1, DES_ENCRYPT);
 		/* remove variant from the first byte of the second half of the key */
 		kek[TDES_KEY_LENGTH_1]^=variant[i];
 	}
 	result = PIN_OK;

	PURGE(kek_ks[0]);
	PURGE(kek_ks[1]);
	return result;
}

File: java/src/org/ilyadubinsky/cfpp/payments/PIN.java
package org.ilyadubinsky.cfpp.payments;

import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.ShortBufferException;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

import org.ilyadubinsky.cfpp.crypto.Constants;
import org.ilyadubinsky.cfpp.crypto.SymmetricAlgorithms;
import org.ilyadubinsky.cfpp.utils.BitOps;
import org.ilyadubinsky.cfpp.utils.IO;

import lombok.extern.java.Log;

@Log

public class PIN {

	public static final int PIN_BLOCK_SIZE_0123_N = 16; // Formats 0 to 3, size in nibbles
	public static final int PIN_BLOCK_SIZE_0123_B = PIN_BLOCK_SIZE_0123_N/2; // Formats 0 to 3, size in bytes

	public static final int PIN_BLOCK_SIZE_4_N = 32; // Format 4, length of a single block in nibbles
	public static final int PIN_BLOCK_SIZE_4_B = PIN_BLOCK_SIZE_4_N / 2; // Format 4 size in bytes

	public static final int PIN_MAX_LENGTH = 12;
	public static final int PAN_MIN_LENGTH = 12;

	public static boolean validAESKeyLength(int length) {
		return (length == (128 >> 3) || length == (192 >> 3) || length == (256 >> 3));
	}

	/**
	 * Prepares a pin block of the given format. The PIN block is written return in
	 * a packed form as a byte array.
	 *
	 * In case of PIN block format 4, two concatenated blocks are returned (they
	 * will have to be CBC encrypted)
	 *
	 * @param format   PIN block format, 0 to 4
	 * @param pin      PIN value, unpacked BCD
	 * @param pan      PAN value, unpacked BCD (can be null depending on the format)
	 * @param uniqueID unique ID value, unpacked BCD (can be null depending on the
	 *                 format)
	 * @return returns the PIN block
	 */
	public static byte[] makePINBlock(int format, byte[] pin, byte[] pan, byte[] uniqueID) {
		if (format < 0 || format > 4 || null == pin || pin.length == 0 || pin.length > PIN_MAX_LENGTH)
			return null;

		if (format != 1 && format != 2 && (null == pan || pan.length == 0 || pan.length <= PAN_MIN_LENGTH))
			return null;

		byte[] unpackedBuffer = new byte[format != 4 ? PIN_BLOCK_SIZE_0123_N : PIN_BLOCK_SIZE_4_N * 2];

		int p = 0;

		/* copy the format and the PIN */
		unpackedBuffer[0] = (byte) (format & 0xF);
		p++;

		System.arraycopy(pin, 0, unpackedBuffer, p, pin.length);

		p += pin.length;

		/*
		 * pad depends on the PIN block format. The logic is: if the format is 0 or 2,
		 * pad with 0xF if the format is 1, and a unique ID is provided, pad with it if
		 * the format is 3, or the format is 1 and there's no unique ID, use random
		 * padding if the format is 4, pad with 0xA the first 8 bytes, then pad with
		 * random bytes
		 */
		switch (format) {

		case 0:
		case 2:
			log.fine(" Block 1: " + IO.printByteArray(BitOps.packBCD(unpackedBuffer, PIN_BLOCK_SIZE_0123_B, false)));
			BitOps.padArray(unpackedBuffer, p, PIN_BLOCK_SIZE_0123_N - p, (byte) 0x0F);
			break;

		case 1:
			if (uniqueID != null) {
				/* pad with unique ID as far as it goes */
				int padLen = PIN_BLOCK_SIZE_0123_N - p > uniqueID.length ? uniqueID.length : PIN_BLOCK_SIZE_0123_N - p;
				System.arraycopy(uniqueID, 0, unpackedBuffer, p, padLen);
				p += padLen;
				log.fine(" Block 1: " + IO.printByteArray(BitOps.packBCD(unpackedBuffer, PIN_BLOCK_SIZE_0123_B, false)));
				if (p == PIN_BLOCK_SIZE_0123_N)
					break;
			}
		case 3:
			/* whatever is left must be padded with a random sequence of bytes */
			byte[] seq = BitOps.randomByteSequence(PIN_BLOCK_SIZE_0123_N - p);
			System.arraycopy(seq, 0, unpackedBuffer, p, PIN_BLOCK_SIZE_0123_N - p);
			break;

		case 4:
			/* pad to length of 16 nibbles with 0xA, then random-pad the rest */
			BitOps.padArray(unpackedBuffer, p, PIN_BLOCK_SIZE_0123_N - p, (byte) 0xA);
			System.arraycopy(BitOps.randomByteSequence(PIN_BLOCK_SIZE_0123_N), 0, unpackedBuffer, PIN_BLOCK_SIZE_0123_N,
					PIN_BLOCK_SIZE_0123_N);
			/* make the second block */
			System.arraycopy(makeFormat4Block2(pan), 0, unpackedBuffer, PIN_BLOCK_SIZE_0123_N * 2,
					PIN_BLOCK_SIZE_0123_N * 2);
			break;
		}

		/* if the format is 0 or 3, the buffer should be XOR-ed with the second block */
		if (format == 0 || format == 3) {
			/* prepare the second block */
			byte[] block2 = new byte[PIN_BLOCK_SIZE_0123_N];
			int panDigitsToCopy = pan.length > 12 ? 12 : pan.length - 1;
			System.arraycopy(pan, pan.length - 1 - panDigitsToCopy, block2, 16 - panDigitsToCopy, panDigitsToCopy);

			log.fine(" Block 2: " + IO.printByteArray(BitOps.packBCD(block2, 8, false)));
			/* XOR the PIN block */
			unpackedBuffer = BitOps.xorArray(unpackedBuffer, block2);

		}
		return BitOps.packBCD(unpackedBuffer, unpackedBuffer.length >> 1, false);
	}

	/**
	 * Utility method to create PIN block format 4 second block
	 * @param pan PAN number, unpacked
	 * @return The unpacked block
	 */
	protected static byte[] makeFormat4Block2(byte[] pan) {

		byte[] block2 = new byte[PIN_BLOCK_SIZE_0123_N * 2];

		int p = 0;
		block2[p++] = (byte) (pan.length - 12);

		System.arraycopy(pan, 0, block2, p, pan.length);

		return block2;
	}

	/**
	 * Encrypts a format 4 block formerly prepared by makePINBlock
	 *
	 * @param key          AES encryption key
	 * @param rawPINBlock4 raw Format 4 PIN block, as returned by makePINBlock
	 * @return encrypted PIN block
	 * @throws NoSuchAlgorithmException
	 * @throws NoSuchPaddingException
	 * @throws InvalidKeyException
	 * @throws IllegalBlockSizeException
	 * @throws BadPaddingException
	 * @throws InvalidAlgorithmParameterException
	 * @throws ShortBufferException
	 */
	public static byte[] encryptPINBlock4(byte[] key, byte[] rawPINBlock4)
			throws NoSuchAlgorithmException, NoSuchPaddingException, InvalidKeyException, IllegalBlockSizeException,
			BadPaddingException, InvalidAlgorithmParameterException, ShortBufferException {
		if (null == key || !validAESKeyLength(key.length) || null == rawPINBlock4
				|| rawPINBlock4.length != PIN_BLOCK_SIZE_4_N)
			return null;

		Cipher c = Cipher.getInstance(Constants.AES_CBC_NO_PADDING);

		SecretKeySpec encKey = new SecretKeySpec(key, Constants.AES_KEY_ALGORITHM);

		IvParameterSpec iv = new IvParameterSpec(new byte[Constants.AES_BLOCK_SIZE_B]);

		c.init(Cipher.ENCRYPT_MODE, encKey, iv);

		/* last block of the encryption output is omitted */

		byte[] retval = new byte[PIN_BLOCK_SIZE_4_B];

		System.arraycopy(c.doFinal(rawPINBlock4), 0, retval, 0, PIN_BLOCK_SIZE_4_B);

		return retval;
	}

	/**
	 * @param key AES decryption key
	 * @param epb Encrypted PIN block
	 * @param pan PAN value
	 * @return Full decrypted PIN block as a packed array
	 * @throws NoSuchAlgorithmException
	 * @throws NoSuchPaddingException
	 * @throws InvalidKeyException
	 * @throws InvalidAlgorithmParameterException
	 * @throws IllegalBlockSizeException
	 * @throws BadPaddingException
	 */
	public static byte[] decryptPINBlock4(byte[] key, byte[] epb, byte[] pan)
			throws NoSuchAlgorithmException, NoSuchPaddingException, InvalidKeyException,
			InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException {
		if (null == key || !validAESKeyLength(key.length) || null == epb || null == pan
				|| epb.length != PIN_BLOCK_SIZE_4_B)
			return null;

		byte[] unpackedBuffer = new byte[PIN_BLOCK_SIZE_4_N * 2];
		/*
		 * to decrypt PIN format 4 PIN block, we take the ciphertext, append the block 2
		 * to it, then use CBC to decipher
		 */
		System.arraycopy(BitOps.unpackBCD(epb), 0, unpackedBuffer, 0, PIN_BLOCK_SIZE_4_N);
		System.arraycopy(makeFormat4Block2(pan), 0, unpackedBuffer, PIN_BLOCK_SIZE_4_N, PIN_BLOCK_SIZE_4_N);
		byte[] packedBuffer = BitOps.packBCD(unpackedBuffer, PIN_BLOCK_SIZE_4_N, false);

		log.fine("Method input: " + IO.printByteArray(epb));
		log.fine("Cipher input: " + IO.printByteArray(packedBuffer));

		/* CBC deciphering */
		Cipher c = Cipher.getInstance(Constants.AES_CBC_NO_PADDING);
		SecretKeySpec encKey = new SecretKeySpec(key, Constants.AES_KEY_ALGORITHM);

		IvParameterSpec iv = new IvParameterSpec(new byte[Constants.AES_BLOCK_SIZE_B]);

		c.init(Cipher.DECRYPT_MODE, encKey, iv);

		return c.doFinal(packedBuffer);
	}

	public static byte[] encryptKeyVariant(  byte[] key, byte [] kek, byte [] variant) throws InvalidKeyException, NoSuchAlgorithmException, NoSuchPaddingException, IllegalBlockSizeException, BadPaddingException {
		if ( null == key || null == kek || null == variant) return null;
		
		/* validate key length */
		if (!SymmetricAlgorithms.isValidTDESKeyLength(key.length) || kek.length != Constants.DES_KEY_SIZE_2_B)
			return null;
		
		/*
		 * the length of the variant must be as big as the multiplicity of the key, i.e.
		 * 1 for single, 2 for double and 3 for triple-length keys
		 */
		if (variant.length <= key.length/Constants.DES_KEY_SIZE_1_B) return null;
		
		log.fine("KEK prior to variants: " + IO.printByteArray(kek));
		
		byte[] buffer = new byte[key.length];
		
		for (int i = 0; i< key.length/Constants.DES_KEY_SIZE_1_B; i++) {
			/* XOR variant byte with the first byte of KEK's second half */
			kek [ Constants.DES_KEY_SIZE_1_B] ^= variant[i];
			
			/* extract the chunk */
			byte[] chunk = new byte[Constants.DES_KEY_SIZE_1_B];
			System.arraycopy(key, i*Constants.DES_KEY_SIZE_1_B, chunk, 0, Constants.DES_KEY_SIZE_1_B);
			
			/* encrypt the key chunk with the variant KEK */
			byte[] encryptedChunk = SymmetricAlgorithms.encryptTDESBlock(chunk, kek);
			System.arraycopy(encryptedChunk, 0, buffer, i*Constants.DES_KEY_SIZE_1_B, Constants.DES_KEY_SIZE_1_B);
			
			/* XOR variant byte with the first byte of KEK's second half again to undo the first XOR */
			kek [ Constants.DES_KEY_SIZE_1_B ] ^= variant[i];
		}

		return buffer;
	}

}
